src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+            groupArray[index++] = query.lastIndex;
+            // This is to stop infinite loop.  Some regular expressions can return 0 length match
+            // which will not advance the lastindex property.  Ex ".*"
+            if (matchArray.index === query.lastIndex) {
+                query.lastIndex++;
+            }
+            if ((index >= matchCountLimit) || (query.lastIndex > searchEndIndex)) {
+                break;
+            }
+        }
+
+        return index / groupArray.groupSize();
+    }
+
+    /**
+     * Determines if the values of specified items in a group array are equal
+     * @private
+     * @param   {GroupArray} groupArray1 First group array
+     * @param   {GroupArray} groupNum1   Item within group array
+     * @param   {GroupArray} groupArray2 Second group array
+     * @param   {GroupArray} groupNum2   Item within 2nd group array
+     * @returns {boolean}  true when values are equal
+     */
+    function _isGroupValuesEqual(groupArray1, groupNum1, groupArray2, groupNum2) {
+        if (_.isEqual(groupArray1.getGroupValue(groupNum1, 0), groupArray2.getGroupValue(groupNum2, 0)) &&
+                _.isEqual(groupArray1.getGroupValue(groupNum1, 1), groupArray2.getGroupValue(groupNum2, 1))) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If the first item of the first group array and the last item of the second group array
+     * are equal, then remove the duplicate from the end of the second group array
+     * @private
+     * @param {GroupArray} firstSearchResults  First search results
+     * @param {GroupArray} secondSearchResults Second search results
+     */
+    function _removeIfDuplicateResultOnEdge(firstSearchResults, secondSearchResults) {
+        if (firstSearchResults.itemCount() === 0 || secondSearchResults.itemCount() === 0) {
+            return;
+        }
+        if (_isGroupValuesEqual(firstSearchResults, 0, secondSearchResults, secondSearchResults.itemCount() - 1)) {
+            secondSearchResults.pop();
+            secondSearchResults.pop();
+        }
+    }
+
+    /**
+     * Creates the regex indexer which finds all matches within supplied text using the search query.
+     * Uses a lookup index to efficiently map regular expression result indexes to position used by Brackets
+     * @param {String} docText       The text to search for matches
+     * @param {Array}  docLineIndex  Array used to map indexes to positions
+     * @param {RegExp} query         A regular expression used to find matches
+     * @param {number} maxResults    The limit of the number of matches to perform
+     * @param {{to: {line: number, ch: number}, from: {line: number, ch: number}} startPosition Start searching at this position
+     * @returns {RegexIndexer} A new instance of the regular expression indexer
+     */
+    function _createRegexIndexer(docText, docLineIndex, query, maxResults, startPosition) {
```

Comment:
Would a name `SearchResults` would be more appropriate for this datatype? Judging by the interface, it would be a more intuitive name for it.